web/tests/copy_and_paste.test.js

```diff
@@ -2,11 +2,47 @@
 
 const {strict: assert} = require("assert");
 
-const {zrequire} = require("./lib/namespace");
+const {zrequire, mock_esm} = require("./lib/namespace");
 const {run_test} = require("./lib/test");
 
 const copy_and_paste = zrequire("copy_and_paste");
 
+mock_esm("../src/stream_data", {
+    slug_to_name() {
+        return "Rome";
+    },
+});
+
+run_test("try_stream_topic_syntax_text", () => {
+    const test_cases = [
+        ["#narrow/stream/4-Rome/topic/old.20FAILED.20EXPORT", true],
+        ["narrow/stream/4-Rome/topic/old.20FAILED.20EXPORT", false],
+        ["#not_narrow/stream/4-Rome/topic/old.20FAILED.20EXPORT", false],
+        ["#narrow/not_stream/4-Rome/topic/old.20FAILED.20EXPORT", false],
+        ["#narrow/stream/4-Rome/not_topic/old.20FAILED.20EXPORT", false],
+        ["#narrow/stream/4-Rome/topic/old.20FAILED.20EXPORT/near/100", false],
+        ["#narrow/stream/4-Rome/", true],
+        ["#narrow/stream/4-Rome/topic", false],
+        ["#narrow/topic/cheese", false],
+        ["#narrow/topic/pizza/stream/Rome", false],
+    ];
+
+    for (const test_case of test_cases) {
+        const result = copy_and_paste.try_stream_topic_syntax_text(
+            window.location.origin + "/" + test_case[0],
+            false, // undo feature isn't needed during tests
+        );
+        const is_transformed = result.startsWith("#**") && result.endsWith("**");
+        assert.equal(is_transformed, test_case[1], "Failed for case: " + test_case[0]);
```

Comment:
These tests are too imprecise; we should be verifying the precise output end-to-end, not just whether we ended up with outer stream/topic mention syntax. (This helps protect against all sorts of encoding bugs).

Reply:
Understood, I am now comparing the exact output.
Also, this feature is prone to the same encoding bugs as discussed in #19873.